lib/std/crypto/pcurves/p256.zig

```diff
@@ -0,0 +1,423 @@
+
+    /// Reject the neutral element.
+    pub fn rejectIdentity(p: P256) IdentityElementError!void {
+        if (p.x.isZero()) {
+            return error.IdentityElement;
+        }
+    }
+
+    /// Create a point from affine coordinates after checking that they match the curve equation.
+    pub fn fromAffineCoordinates(x: Fe, y: Fe) EncodingError!P256 {
+        const x3AxB = x.sq().mul(x).sub(x).sub(x).sub(x).add(B);
+        const yy = y.sq();
+        if (!x3AxB.equivalent(yy)) {
+            return error.InvalidEncoding;
+        }
+        const p: P256 = .{ .x = x, .y = y, .z = Fe.one };
+        return p;
+    }
+
+    /// Create a point from serialized affine coordinates.
+    pub fn fromSerializedAffineCoordinates(xs: [32]u8, ys: [32]u8, endian: builtin.Endian) (NonCanonicalError || EncodingError)!P256 {
+        const x = try Fe.fromBytes(xs, endian);
+        const y = try Fe.fromBytes(ys, endian);
+        return fromAffineCoordinates(x, y);
+    }
+
+    /// Recover the Y coordinate from the X coordinate.
+    pub fn recoverY(x: Fe, is_odd: bool) NotSquareError!Fe {
+        const x3AxB = x.sq().mul(x).sub(x).sub(x).sub(x).add(B);
+        var y = try x3AxB.sqrt();
+        const yn = y.neg();
+        y.cMov(yn, @boolToInt(is_odd) ^ @boolToInt(y.isOdd()));
+        return y;
+    }
+
+    /// Deserialize a SEC1-encoded point.
+    pub fn fromSec1(s: []const u8) (EncodingError || NotSquareError || NonCanonicalError)!P256 {
+        if (s.len < 1) return error.InvalidEncoding;
+        const encoding_type = s[0];
+        const encoded = s[1..];
+        switch (encoding_type) {
+            0 => {
+                if (encoded.len != 0) return error.InvalidEncoding;
+                return P256.identityElement;
+            },
+            2, 3 => {
+                if (encoded.len != 32) return error.InvalidEncoding;
+                const x = try Fe.fromBytes(encoded[0..32].*, .Big);
+                const y_is_odd = (encoding_type == 3);
+                const y = try recoverY(x, y_is_odd);
+                return P256{ .x = x, .y = y };
+            },
+            4 => {
+                if (encoded.len != 64) return error.InvalidEncoding;
+                const x = try Fe.fromBytes(encoded[0..32].*, .Big);
+                const y = try Fe.fromBytes(encoded[32..64].*, .Big);
+                return P256.fromAffineCoordinates(x, y);
+            },
+            else => return error.InvalidEncoding,
+        }
+    }
+
+    /// Serialize a point using the compressed SEC-1 format.
+    pub fn toCompressedSec1(p: P256) [33]u8 {
+        var out: [33]u8 = undefined;
+        const xy = p.affineCoordinates();
+        out[0] = if (xy.y.isOdd()) 3 else 2;
+        mem.copy(u8, out[1..], &xy.x.toBytes(.Big));
+        return out;
+    }
+
+    /// Serialize a point using the uncompressed SEC-1 format.
+    pub fn toUncompressedSec1(p: P256) [65]u8 {
```

Comment:
since we have `fromSec1`
maybe this could be renamed `toSec1`

being uncompressed is the default state, it's only worth having for `toCompressedSec1` IMO

Reply:
`fromSec1()` can parse compressed and uncompressed representations, as this information is present.

A `toSec1()` function would would have to make a choice (or get an ugly API) since the output sizes of both representations differ.

Compressed vs uncompressed is a never ending debate (former is smaller and more secure, later is more common), so maybe we should stay neutral and give no advantage to either wire format.

Reply:
> or get an ugly API

(p: P256, c: bool) => [65]u8 | [33]u8?
Yeah that smells.

> we should stay neutral and give no advantage to either wire format

👍